### PR TITLE
Igvc net

### PR DESCRIPTION
**Pull request by jzheng84**
This PR does the following:
* Add the train/evaluation pipeline for training a model for line detection in pytorch.
* A ROS node that runs inference for line detection using a pytorch model.
* Moves old line detection code into respective "legacy" files.

### CODE REVIEW

File: igvc_perception/src/train_eval/evaluate.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    if anno_dir:
+        avg_accuracy /= len(img_names)
+        recalls = []
+        precisions = []
+        keys = avg_recalls.keys()
+        keys.sort()
+        for key in keys:
+            avg_recalls[key] /= len(img_names)
+            avg_precisions[key] /= len(img_names)
```
Comment (Alescontrela) [inline igvc_perception/src/train_eval/evaluate.py]: `avg_recalls` and `avg_precisions` could be quite easily fused to calculate the F1 score as the harmonic mean of precision and recall. [source](https://en.wikipedia.org/wiki/F1_score#Definition)

File: igvc_perception/src/train_eval/models/model.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+import torch
```
Comment (Alescontrela) [inline igvc_perception/src/train_eval/models/model.py]: I would include a comment at the top of this file linking to the UNet paper. [link](https://arxiv.org/abs/1505.04597)

File: igvc_perception/src/train_eval/train.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    model = torch.load(args.load_model)
+elif args.test:
+    print('Missing model file for evaluating test set.')
+    exit()
+else:
+    model = models.model.UNet(args.im_size, args.kernel_size)
+
+# Datasets and dataloaders.
+if not args.test:
+    train_dataset = IGVCDataset(train_txt, im_size=args.im_size, split='train', transform=transform, val_samples=args.val_samples)
```
Comment (Alescontrela) [inline igvc_perception/src/train_eval/train.py]: This is a really nice way to handle image datasets. Good to know!

File: igvc_perception/launch/vision.launch [hunk 4]
```diff
@@ -1,16 +1,22 @@
...
 <launch>
+
     <node name="cnn" pkg="igvc_perception" type="cnn.py" output="screen" />
 
-    <param name="/image_topic" value="/usb_cam_center/image_raw" />
-    <param name="/image_resize_width" value="300" />
-    <param name="/image_resize_height" value="225" />
+    <rosparam param="camera_names">["usb_cam_center"]</rosparam>
+
+    <!-- These sizes must match the model input dimensions. -->
```
[the remark this answers is not in the document]
Comment (JasonGibson274) [reply igvc_perception/launch/vision.launch]: I think it is fine if you plan on running with multiple sized models, or not having to edit source code to change the model.

File: igvc_perception/launch/vision_legacy.launch [hunk 5]
```diff
@@ -0,0 +1,16 @@
+<!-- vision.launch -->
```
[the remark this answers is not in the document]
Comment (JasonGibson274) [reply igvc_perception/launch/vision_legacy.launch]: Yes, unless we plan on running both I would delete this one.

File: igvc_perception/src/cnn/cnn.py [hunk 6]
```diff
@@ -1,139 +1,175 @@
...
+        self.line_thresh = line_thresh
+        
+        # Setup camera model and subsciber for each camera.
+        self.camera_models = {}
+        self.cam_transform_rotation_matrices = {}
+        self.cam_transform_translations = {}
+        self.world_point_arrays = {}
+        self.subscribers = []
+
+        self.cam_frame_names = {'usb_cam_center': '/optical_cam_center',
```
Comment (JasonGibson274) [inline igvc_perception/src/cnn/cnn.py]: I would just write a small function that assumes a standard naming convention. The camera names are a parameter but only one that can take up to three values.

File: igvc_perception/launch/vision_legacy.launch (hunk 5 above)
[the remark this answers is not in the document]
Comment (jzheng84) [reply igvc_perception/launch/vision_legacy.launch]: removed

File: igvc_perception/src/cnn/cnn.py [hunk 7]
```diff
@@ -1,139 +1,175 @@
...
+
+            # adjust the projection matrix.
+            P = camera_info.P
+            camera_info.P = (P[0]*waf,        0.,  P[2]*waf,  0.,
+                                   0.,  P[5]*haf,  P[6]*haf,  0.,
+                                   0.,        0.,        1.,  0.)
+
+            self.camera_models[camera_name] = PinholeCameraModel()
+            self.camera_models[camera_name].fromCameraInfo(camera_info)
+            print camera_info
```
[the remark this answers is not in the document]
Comment (jzheng84) [reply igvc_perception/src/cnn/cnn.py]: done

File: igvc_perception/src/cnn/cnn.py [hunk 8]
```diff
@@ -1,139 +1,175 @@
...
+
+            # Get transform between camera and base_footprint.
+            transform_listener = ros_tf.TransformListener()
+
+            cam_frame_name = self.cam_frame_names[camera_name]
+            transform_listener.waitForTransform('/base_footprint', cam_frame_name, rospy.Time(0), rospy.Duration(5.0))
+            cam_transform_translation, cam_transform_rotation = transform_listener.lookupTransform('/base_footprint', cam_frame_name, rospy.Time(0))
+            self.cam_transform_rotation_matrices[camera_name] = ros_tf.transformations.quaternion_matrix(cam_transform_rotation)[:-1,:-1]
+            self.cam_transform_translations[camera_name] = np.asarray(cam_transform_translation)
+            print self.cam_transform_translations[camera_name]
```
[the remark this answers is not in the document]
Comment (jzheng84) [reply igvc_perception/src/cnn/cnn.py]: changes to logging, leaving it there just to make sure cameras are launched properly

File: igvc_perception/src/cnn/cnn_legacy.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (jzheng84) [reply igvc_perception/src/cnn/cnn_legacy.py]: removed

File: igvc_perception/src/train_eval/cfg/igvc.cfg [hunk 10]
```diff
@@ -0,0 +1,3 @@
+train = /media/justin_zheng/OS/Data/igvc/3-2/train.txt
```
[the remark this answers is not in the document]
Comment (jzheng84) [reply igvc_perception/src/train_eval/cfg/igvc.cfg]: done

File: igvc_perception/src/train_eval/evaluate.py (hunk 1 above)
Comment (jzheng84) [reply igvc_perception/src/train_eval/evaluate.py]: added

File: igvc_perception/src/train_eval/models/model.py (hunk 2 above)
Comment (jzheng84) [reply igvc_perception/src/train_eval/models/model.py]: done

File: igvc_perception/src/train_eval/train.py (hunk 3 above)
Comment (jzheng84) [reply igvc_perception/src/train_eval/train.py]: thanks :^)

File: igvc_perception/src/cnn/cnn.py (hunk 6 above)
Comment (jzheng84) [reply igvc_perception/src/cnn/cnn.py]: done, just replaced using simple "replace" function

File: igvc_perception/src/train_eval/train.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+        if args.cuda:
+            images, targets = images.cuda(), targets.cuda()
+
+        optimizer.zero_grad()
+        outputs = model(images)
+        loss = criterion(outputs, targets)
+        loss.backward()
+        optimizer.step()
+
+        if args.vis and batch_idx % args.log_interval == 0 and images.shape[0] == 1:
```
Comment (Alescontrela) [inline igvc_perception/src/train_eval/train.py]: I find that, if I make `batch_size` > 1, i'm no longer able to visualize the model's output

Comment (Alescontrela) [reply igvc_perception/src/train_eval/train.py]: I was able to "fix" this by removing the `images.shape[0] == 1` condition. Regardless of the number of batches, `outputs.cpu().data.numpy().shape` always has the same format: [b,n_c,w,h], where b is the number of batches:

```python
if args.vis and (batch_idx % args.log_interval == 0):

            cv2.imshow('output: ', np.hstack(tuple([outputs.cpu().data.numpy()[i][0] for i in range(outputs.cpu().data.numpy().shape[0])])))
            cv2.imshow('target: ', np.hstack(tuple([targets.cpu().data.numpy()[i][0] for i in range(outputs.cpu().data.numpy().shape[0])])))
            cv2.waitKey(10)
```

Comment (Alescontrela) [reply igvc_perception/src/train_eval/train.py]: Resolving as this has no effect on functionality

**Pull request merged**